### PR TITLE
issue #7: add initial support for istio authorization policies, WIP

### PR DESCRIPTION
**Pull request by adisos**

### CODE REVIEW

File: network-config-analyzer/NetworkConfigQuery.py [hunk 1]
```diff
@@ -269,11 +275,11 @@ def other_policy_containing_deny(self, self_policy, config_with_self_policy):
...
                         continue  # no way to prevent a pod from communicating with itself
-                    _, _, self_deny_conns = config_with_self_policy.allowed_connections(pod1, pod2)
-                    _, _, other_deny_conns = config_with_other_policy.allowed_connections(pod1, pod2)
-                    if not self_deny_conns:
+                    _, _, _, self_conns = config_with_self_policy.allowed_connections(pod1, pod2)
+                    _, _, _, other_conns = config_with_other_policy.allowed_connections(pod1, pod2)
+                    if not self_conns:
                         continue
-                    if not self_deny_conns.contained_in(other_deny_conns):
+                    if not self_conns.contained_in(other_conns):
```
Comment (zivnevo) [inline network-config-analyzer/NetworkConfigQuery.py:282]: why remove the deny?

File: network-config-analyzer/NetworkConfigQuery.py [hunk 2]
```diff
@@ -475,6 +488,17 @@ def exec(self):
             res.output_explanation = fw_rules.get_fw_rules_in_required_format()
         return res
 
+    def filter_istio_edge(self, peer2, conns):
```
Comment (zivnevo) [inline network-config-analyzer/NetworkConfigQuery.py:491]: should be a static method

File: network-config-analyzer/IstioNetworkPolicy.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                emptiness_explanation.append(emptiness)
+                empty_ingress_rules.add(rule_index)
+
+        return emptiness_explanation, empty_ingress_rules, set()
+
+    def clone_without_rule(self, rule_to_exclude, ingress_rule):
+        """
+        Makes a copy of 'self' without a given policy rule
+        :param IstioPolicyRule rule_to_exclude: The one rule not to include in the copy
+        :param bool ingress_rule: Whether the rule is an ingress or egress rule
```
Comment (zivnevo) [inline network-config-analyzer/IstioNetworkPolicy.py:131]:
```suggestion
    def clone_without_rule(self, rule_to_exclude, _ingress_rule):
        """
        Makes a copy of 'self' without a given policy rule
        :param IstioPolicyRule rule_to_exclude: The one rule not to include in the copy
        :param bool _ingress_rule: Whether the rule is an ingress or egress rule
```
The underscore marks the variable is not used

File: network-config-analyzer/IstioNetworkPolicy.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        :param str config_name: (optional) the name of the NetworkConfig object
+        :return: A list of strings describing the emptiness + two sets of indexes of empty ingress/egress rules
+        :rtype: list[str], set, set
+        """
+        emptiness_explanation = []
+        empty_ingress_rules = set()
+        full_name = self.full_name(config_name)
+        for rule_index, ingress_rule in enumerate(self.ingress_rules, start=1):
+            if not ingress_rule.peer_set:
+                emptiness = f'Ingress rule no. {rule_index} in AuthorizationPolicy {full_name} does not select any pods'
```
Comment (zivnevo) [inline network-config-analyzer/IstioNetworkPolicy.py:121]: Should we leave the `Ingress` here? It is kind of obvious

File: network-config-analyzer/IstioNetworkPolicy.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        allowed_conns = ConnectionSet()
+        denied_conns = ConnectionSet()
+
+        rules = self.ingress_rules
+        collected_conns = allowed_conns if self.action == IstioNetworkPolicy.ActionType.Allow else denied_conns
+        for rule in rules:
+            if from_peer in rule.peer_set:
+                rule_conns = rule.connections
+                collected_conns |= rule_conns
```
Comment (zivnevo) [inline network-config-analyzer/IstioNetworkPolicy.py:93]:
```suggestion
        collected_conns = allowed_conns if self.action == IstioNetworkPolicy.ActionType.Allow else denied_conns
        for rule in self.ingress_rules:
            if from_peer in rule.peer_set:
                collected_conns |= rule.connections
```

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
+#
+# Copyright 2020- IBM Inc. All rights reserved
+# SPDX-License-Identifier: Apache2.0
+#
+
+from GenericYamlParser import GenericYamlParser
+from IstioNetworkPolicy import IstioNetworkPolicy, IstioPolicyRule
+from Peer import IpBlock
+from Peer import PeerSet
```
Comment (zivnevo) [inline network-config-analyzer/IstioPolicyYamlParser.py:9]:
```suggestion
from Peer import IpBlock, PeerSet
```

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+        if self.policy.get('kind') != 'AuthorizationPolicy':
+            return None  # Not a AuthorizationPolicy object
+        api_version = self.policy.get('apiVersion', self.policy.get('api_version', ''))
+        if not api_version:
+            self.syntax_error('An object with no specified apiVersion', self.policy)
+        if 'istio' not in api_version:
+            return None  # apiVersion is not properly set
+        if api_version not in ['security.istio.io/v1beta1']:
+            raise Exception('Unsupported apiVersion: ' + api_version)
+        self.check_fields_validity(self.policy, 'networkPolicy', {'kind': 1, 'metadata': 1, 'spec': 1,
```
Comment (zivnevo) [inline network-config-analyzer/IstioPolicyYamlParser.py:341]:
```suggestion
        self.check_fields_validity(self.policy, 'AuthorizationPolicy', {'kind': 1, 'metadata': 1, 'spec': 1,
```

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+                                                                  'apiVersion': 0, 'api_version': 0})
+        if 'name' not in self.policy['metadata']:
+            self.syntax_error('NetworkPolicy has no name', self.policy)
```
Comment (zivnevo) [inline network-config-analyzer/IstioPolicyYamlParser.py:344]:
```suggestion
            self.syntax_error('AuthorizationPolicy has no name', self.policy)
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
Something to start with (I'm not done with reviewing the parser).

One general comment:
I think it's better to give testcases (folders, policy files and scheme files) more meaningful names that hint what they test. `testcase2-policy-3.yaml` doesn't tell much. No need to change now, but maybe from now on.

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+        :return: ConnectionSet object with relevant TCP connections
+        """
+        connections = ConnectionSet()
+        if ports_list is None:  # If not set, any port is allowed.
+            connections = self.get_all_ports_connection()
+        else:
+            for port in ports_list:
+                connections |= self.parse_port(port)
+        not_ports_list = [] if not_ports_list is None else not_ports_list
+        for port in not_ports_list:
```
Comment (zivnevo) [inline network-config-analyzer/IstioPolicyYamlParser.py:154]:
```suggestion
        for port in not_ports_list or []:
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
Done reviewing the parser. A few more comments

File: tests/fw_rules_tests/policies/istio-test1-policy1.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    - source:
+        ipBlocks: ["1.2.3.0/24", "2.2.2.2"]
+    to:
+    - operation:
+        ports: ["26257", "30"]
+    when:
+    - key: destination.port
+      values: ["26257"]
+    #- key: source.namespace
+    #  values: ["default"]
```
[the remark this answers is not in the document]
Comment (adisos) [reply tests/fw_rules_tests/policies/istio-test1-policy1.yaml]: done

File: tests/fw_rules_tests/policies/istio-test1-policy1.yaml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+---
+apiVersion: security.istio.io/v1beta1
+kind: AuthorizationPolicy
+metadata:
+  name: test-deny-nothing
+  namespace: vendor-system
+spec:
+  action: DENY #ALLOW
+  #rules:
+  #  - {}
```
[the remark this answers is not in the document]
Comment (adisos) [reply tests/fw_rules_tests/policies/istio-test1-policy1.yaml]: done

File: tests/fw_rules_tests/policies/istio-test1-policy2.yaml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        ports: ["26257", "30", "40", "50"]
+        notPorts: ["40"]
+    when:
+    - key: destination.port
+      notValues: ["26257"]
+      #values: ["30"]
+    #- key: source.namespace
+    #  values: ["default"]
+    #- key: source.ip
+    #  notValues: ["2.2.2.2"]
```
[the remark this answers is not in the document]
Comment (adisos) [reply tests/fw_rules_tests/policies/istio-test1-policy2.yaml]: done

File: tests/istio_testcases/example_policies/testcase1/testcase1-deny-all-test-policy.yaml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+apiVersion: security.istio.io/v1beta1
+kind: AuthorizationPolicy
+metadata:
+ name: allow-nothing
+ namespace: default
+spec:
+  {}
```
[the remark this answers is not in the document]
Comment (adisos) [reply tests/istio_testcases/example_policies/testcase1/testcase1-deny-all-test-policy.yaml]: renamed the policy file to "testcase1-allow-nothing-policy-2.yaml" , and added to the scheme file.

File: network-config-analyzer/NetworkPolicy.py [hunk 14]
```diff
@@ -23,6 +24,7 @@ class PolicyType(Enum):
         CalicoGlobalNetworkPolicy = 3
         CalicoProfile = 4
         List = 5
+        IstioAuthorizationPolicy = 6
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/NetworkPolicy.py]: done

File: network-config-analyzer/NetworkConfig.py [hunk 15]
```diff
@@ -314,13 +324,39 @@ def _get_profile_conns(self, from_peer, to_peer, is_ingress):
...
+        TCP_conns.add_connections('TCP', PortSetPair(PortSet(True), PortSet(True)))
+        return ConnectionSet(True) - TCP_conns
+
     def _allowed_xgress_conns(self, from_peer, to_peer, is_ingress):
+        """
+        get allowed and denied ingress/egress connections between from_peer and to_peer,
+        considering all config's policies (and defaults)
+        :param from_peer: the source peer
+        :param to_peer: the dest peer
+        :param is_ingress: flag to indicate if should retuen ingress conections or gress connections
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/NetworkConfig.py]: done

File: network-config-analyzer/NetworkConfig.py [hunk 16]
```diff
@@ -314,13 +324,39 @@ def _get_profile_conns(self, from_peer, to_peer, is_ingress):
...
         policy_captured = False
+        has_allow_policies_for_target = False
         for policy in self.sorted_policies:
             policy_conns = policy.allowed_connections(from_peer, to_peer, is_ingress)
+            if policy_conns.captured and self._get_policy_type(policy) == NetworkConfig.ConfigType.Istio:
+                if policy.action == IstioNetworkPolicy.ActionType.Allow:
+                    has_allow_policies_for_target = True
             assert isinstance(policy_conns, PolicyConnections)
             if policy_conns.captured:
                 policy_captured = True
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/NetworkConfig.py]: done

File: network-config-analyzer/NetworkConfig.py [hunk 17]
```diff
@@ -314,13 +324,39 @@ def _get_profile_conns(self, from_peer, to_peer, is_ingress):
             raise Exception(peer.full_name() + ' refers to a non-existing profile ' + profile_name)
         return profile.allowed_connections(from_peer, to_peer, is_ingress)
 
+    @staticmethod
+    # TODO: move to some other class?
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/NetworkConfig.py]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 18]
```diff
@@ -205,6 +205,9 @@ class SanityQuery(NetworkConfigQuery):
     """
 
     def has_conflicting_policies_with_same_order(self):
+        # TODO: should skip istio policies here?
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/NetworkConfigQuery.py]: done

File: network-config-analyzer/NetworkConfigQuery.py (hunk 1 above)
Comment (adisos) [reply network-config-analyzer/NetworkConfigQuery.py]: fixed (removed by mistake)

File: network-config-analyzer/NetworkConfigQuery.py (hunk 2 above)
Comment (adisos) [reply network-config-analyzer/NetworkConfigQuery.py]: done

File: network-config-analyzer/IstioNetworkPolicy.py (hunk 3 above)
Comment (adisos) [reply network-config-analyzer/IstioNetworkPolicy.py]: done

File: network-config-analyzer/IstioNetworkPolicy.py (hunk 4 above)
Comment (adisos) [reply network-config-analyzer/IstioNetworkPolicy.py]: done

File: network-config-analyzer/IstioNetworkPolicy.py (hunk 5 above)
Comment (adisos) [reply network-config-analyzer/IstioNetworkPolicy.py]: done

File: network-config-analyzer/IstioPolicyYamlParser.py (hunk 6 above)
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done

File: network-config-analyzer/IstioPolicyYamlParser.py (hunk 7 above)
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done

File: network-config-analyzer/IstioPolicyYamlParser.py (hunk 8 above)
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+        :param dict label_selector: The element to parse
+        :return: A PeerSet containing all the pods captured by this selection
+        :rtype: Peer.PeerSet
+        """
+        if label_selector is None:
+            return PeerSet()  # A None value means the selector selects nothing
+        if not label_selector:
+            return self.peer_container.get_all_peers_group()  # An empty value means the selector selects everything
+
+        allowed_elements = {'matchLabels': 0, 'match_labels': 0}
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done (added match_labels by mistake)

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+        self.check_fields_validity(label_selector, 'authorization policy WorkloadSelector', allowed_elements)
+
+        res = self.peer_container.get_all_peers_group()
+        match_labels = label_selector.get('matchLabels', label_selector.get('match_labels'))
+        if match_labels:
+            keys_set = set()
+            for key, val in match_labels.items():
+                res &= self.peer_container.get_peers_with_label(key, [val])
+                keys_set.add(key)
+            self.allowed_labels.add(':'.join(keys_set))
```
[the remark this answers is not in the document]
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done

File: network-config-analyzer/IstioPolicyYamlParser.py (hunk 9 above)
Comment (adisos) [reply network-config-analyzer/IstioPolicyYamlParser.py]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 21]
```diff
@@ -1006,10 +1031,11 @@ def exec(self, only_captured):
...
                 if peer1 == peer2:
                     continue
-                captured1, conns1, _ = self.config1.allowed_connections(peer1, peer2)
-                if only_captured and not captured1:
+                conns1_all, captured1_flag, conns1_captured, _ = self.config1.allowed_connections(peer1, peer2)
+                if only_captured and not captured1_flag:
                     continue
-                _, conns2, _ = self.config2.allowed_connections(peer1, peer2)
+                conns1 = conns1_captured if only_captured else conns1_all
+                conns2, _, _, _, = self.config2.allowed_connections(peer1, peer2)
```
Comment (zivnevo) [inline network-config-analyzer/NetworkConfigQuery.py:1038]:
```suggestion
                conns2, _, _, _ = self.config2.allowed_connections(peer1, peer2)
```
A redundant comma (not sure what's the effect it has)

**Review by zivnevo — CHANGES_REQUESTED** [review]
One last tiny fix and we are set

**Review by zivnevo — APPROVED** [review]
ok to merge

**Pull request merged**